src/peft/tuners/vblora/layer.py

```diff
@@ -0,0 +1,263 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import warnings
+from typing import List, Optional
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from transformers.pytorch_utils import Conv1D
+
+from peft.tuners.tuners_utils import BaseTunerLayer, check_adapters_to_merge
+from peft.utils.other import transpose
+
+
+class VBLoRALayer(BaseTunerLayer):
+    # List all names of layers that may contain adapter weights
+    adapter_layer_names = ("vblora_logits_A", "vblora_logits_B", "vblora_vector_bank")
+
+    def __init__(self, base_layer: nn.Module, **kwargs):
+        self.base_layer = base_layer
+        self.r = {}
+        self.topk = {}
+        self.vblora_dropout = nn.ModuleDict({})
+
+        # For storing vector scale
+        self.vblora_logits_A = nn.ParameterDict({})
+        self.vblora_logits_B = nn.ParameterDict({})
+        self.vblora_vector_bank = nn.ParameterDict({})
+
+        # Mark the weight as unmerged
+        self._disable_adapters = False
+        self.merged_adapters = []
+
+        base_layer = self.get_base_layer()
+        if isinstance(base_layer, nn.Linear):
+            in_features, out_features = base_layer.in_features, base_layer.out_features
+        elif isinstance(base_layer, Conv1D):
+            in_features, out_features = (
+                base_layer.weight.ds_shape if hasattr(base_layer.weight, "ds_shape") else base_layer.weight.shape
+            )
+
+        self.in_features = in_features
+        self.out_features = out_features
+        self.kwargs = kwargs
+
+    @property
+    def merged(self) -> bool:
+        return bool(self.merged_adapters)
+
+    def update_layer(
+        self,
+        adapter_name,
+        vblora_vector_bank,
+        r,
+        topk,
+        num_vectors,
+        vector_length,
+        vblora_dropout,
+        init_logits_std,
+    ):
+        if r <= 0:
+            raise ValueError(f"`r` {r} should be a positive integer value")
+        if topk <= 0:
+            raise ValueError(f"`topk` {topk} should be a positive integer value")
+
+        if self.in_features % vector_length != 0:
+            raise ValueError(f"`in_features` {self.in_features} must be divisible by `vector_length` {vector_length}")
+        if self.out_features % vector_length != 0:
+            raise ValueError(
+                f"`out_features` {self.out_features} must be divisible by `vector_length` {vector_length}"
+            )
```

Comment:
It would be great to have tests for this. They can go to `test_initialization.py`.

Reply:
Agree. will add tests for this.

Reply:
Added a new test.